openapi/resource-server.yaml

```diff
@@ -959,7 +967,6 @@ components:
           metadata:
             description: Thank you for your purchase at ShoeShop!
             externalRef: INV2022-8943756
-      additionalProperties: false
```

Comment:
should this be deleted? I feel like we had a discussion around `additionalProperties`, but I don't remember it

Reply:
Actually no, this doesn't need to be deleted. I returned it.  It's left over from when I made `outgoing-payment` an `allOf` with the `outgoing-payment` and the new grant spent amounts. But then I switched to inline everything in a new `outgoing-payment-with-spent-amounts` model. That is what you're thinking of and is described in this comment: https://github.com/interledger/open-payments/pull/474#discussion_r1607102838 And this slack thread: https://interledgerfoundation.slack.com/archives/C06KZCKEEKW/p1716390398470419

Basically, using `allOf` to extend models (in this case extending `outgoing-payment` with the grant spent amounts) requires allowing additional properties. I went back and forth on if I wanted to do it this way. It's more convenient for writing models and removes duplication. We do this elsewhere and its in the open api docs as a way to extend models. However, I didn't really intend to signal that additional properties are actually allowed beyond that (in the docs, in validation, or anything else that consumes the spec). Especially given the recent confusion about the `key: any` in the wallet address docs (which is driven by this).

There is a separate argument to be made (and has been made by some at Fynbos IIRC) that we shouldn't disallow `additionalProperties` by default. There may be a good case there, but to this point we've defaulted to not allowing so that's why I'm hesitant to signal they are allowed merely so we can use `allOf` to extend types in the yaml file instead of inlining.

Reply:
I guess the potential counterpoint to my reasoning is that allowing `additionalProperties` isn’t actually a bad thing - just means we have no opinion on additional properties. In which case we can also enjoy using `allOf` to extend `outgoing-payment`.

Reply:
What is the default behaviour if we don't include `additionalProperties` at all? Does it allow for them? I don't mind leaving it out, as long as it doesn't show `key: any` in the site, since that might be confusing to see on something like the outgoing payment creation API.

We can keep this for now and then at some point just go through the whole spec and remove it, if we choose to do so.

Reply:
> What is the default behaviour if we don't include additionalProperties at all?

The short answer is it defaults to additionalProperties: true, and this field is what drives the key: any in the docs. So I think my current changes jive with how we've (mostly) been doing things and conforms to your expectations for the docs which is something I considered myself.

This is veering into a larger conversation, but zooming out a bit, I'm starting to think we shouldn't specify additionalProperties: false unless we actually have a reason to signal that there shouldn't be any additional properties. Which I don't think we have for any of the existing models. Letting additionalProperties default to true just means we have no opinion on if implementations return additional stuff. But currently we are saying "there should never be any additional properties returned by a quote," for example. Why? I don't really have a defense for this requirement. I can't anticipate why someone would want to do that, but I also don't have a specific reason for why they shouldn't do it. If we are removing some capability I think we need to have a good reason.